pkg/activator/net/revision_backends.go

```diff
@@ -0,0 +1,418 @@
+
+	var probeGroup errgroup.Group
+	healthStatesCh := make(chan destHealth, len(rw.dests))
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthStates[dest]; ok && curHealthy {
+			healthStatesCh <- destHealth{dest, true}
+			continue
+		}
+
+		pDest := dest
+		probeGroup.Go(func() error {
+			ok, err := rw.probe(ctx, pDest)
+			healthStatesCh <- destHealth{pDest, ok}
+			return err
+		})
+	}
+
+	if err := probeGroup.Wait(); err != nil {
+		return nil, err
+	}
+	close(healthStatesCh)
+
+	healthStates := make(map[string]bool, len(rw.dests))
+	for dh := range healthStatesCh {
+		healthStates[dh.dest] = dh.health
+	}
+
+	return healthStates, nil
+}
+
+// checkDests performs probing and potentially sends a dests update. It is
+// assumed this method is not called concurrently.
+func (rw *revisionWatcher) checkDests() {
+	if !rw.revisionReady() {
+		if rw.clusterIPHealthy {
+			// we have a healthy clusterIP but revision is not ready. We must have scaled down.
+			rw.clusterIPHealthy = false
+
+			// Send update that were now inactive
+			rw.updateCh <- &RevisionDestsUpdate{Rev: rw.rev}
+		}
+
+		// We know this revision cannot be healthy so short circuit
+		return
+	}
+
+	if rw.clusterIPHealthy {
+		// cluster IP is healthy and we havent scaled down, short circuit
+		return
+	}
+
+	// First check the clusterIP
+	svc, err := rw.getK8sPrivateService()
+	if err != nil {
+		rw.logger.Errorw("Failed to lookup private service for revision", zap.Error(err))
+		return
+	}
+
+	// If clusterIP is healthy send this update and we are done
+	if ok, dest, err := rw.probeClusterIP(svc); err != nil {
+		rw.logger.Errorw(fmt.Sprintf("Failed to probe clusterIP %s/%s", svc.Namespace, svc.Name), zap.Error(err))
+	} else if ok {
+		rw.clusterIPHealthy = true
+		rw.healthStates = make(map[string]bool)
+		rw.updateCh <- &RevisionDestsUpdate{Rev: rw.rev, ClusterIPDest: dest, ReadyAddressCount: len(rw.dests)}
+		return
+	}
+
+	healthStates, err := rw.probePodIPs()
+	if err != nil {
+		rw.logger.Errorw("Failed probing", zap.Error(err))
+		// We dont want to return here as an error still affects health states
+	}
+
+	rw.logger.Debugf("Done probing, got healthStates %v", healthStates)
+	if !reflect.DeepEqual(rw.healthStates, healthStates) {
+		rw.healthStates = healthStates
+		destsArr := filterHealthyDests(healthStates)
+		rw.updateCh <- &RevisionDestsUpdate{
+			Rev:               rw.rev,
+			Dests:             destsArr,
+			ReadyAddressCount: len(rw.dests),
+		}
+	}
+}
+
+func (rw *revisionWatcher) runWithTickCh(tickCh <-chan time.Time) {
+	for {
+		select {
+		case dests, ok := <-rw.destsChan:
+			if !ok {
+				// shutdown
+				return
+			}
+			rw.dests = dests
+		case <-tickCh:
+		}
+		rw.checkDests()
+	}
+}
+
+func (rw *revisionWatcher) run(probeFrequency time.Duration) {
+	ticker := time.NewTicker(probeFrequency)
+	defer ticker.Stop()
+
+	rw.runWithTickCh(ticker.C)
+}
+
+type revisionWatcherCh struct {
```

Comment:
this struct seems unnecessary since `ch` will always point to  `revisionWatcher.destsChan`

Reply:
Its for the readonly type safety of ch in revisionWatcher. I'm not sure its the best answer, but I'd need to make revisionWatcher.destsChan writable so we can close it when we need to.